FreeRTOS/FreeRTOS-Plus-TCP
# DHCP network buffers

Description
-----------
In FreeRTOS_DHCP.c, a network buffer is obtained in a blocking way. This is no problem because during the DHCP negociation, the device will drop almost all incoming packets. But if it runs out of network buffers, the DHCP process could stall, thus stalling the IP-task.

In this PR, `pxGetNetworkBufferWithDescriptor()` will be called in a polling way ( with a time-out of zero ). When it fails, a new attempt will be done after 250 msec, a time-out defined by this macro:
~~~c
#define dhcpINITIAL_TIMER_PERIOD    ( pdMS_TO_TICKS( 250U ) )
~~~


Test Steps
-----------
I tested it by simulating a shortage of packages, using the following source of randomness:
~~~
static BaseType_t prvDoCreateMessage( void )
{
    uint32_t ulValue;

    xApplicationGetRandomNumber( &( ulValue ) );
    /* Put in the range 0..99 */
    ulValue = ulValue % 100U;
    if (EP_DHCPData.eDHCPState == eSendDHCPRequest)
    {
        return ulValue < 25;
    }
    else
    {
        return ulValue < 5;
    }
}
~~~

and have `pxGetNetworkBufferWithDescriptor()` fail if `prvDoCreateMessage()` returns a zero.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
